### ISSUE TITLE
Reasoning about definedness of Map-typed collection cells

### ISSUE THREAD

**Issue by daejunpark** [kframework/kore#1669]
This is somehow related to a K-frontend issue (https://github.com/kframework/k/issues/1157), but a slightly different issue of the haskell backend.

Let me consider the same example used in the K-frontend issue, that is:

a.k:
```
module A-SYNTAX
imports INT-SYNTAX
syntax Cmds ::= List{Cmd,""}
syntax Cmd ::= add(Int, Int)
rule C:Cmd Cs:Cmds => C ~> Cs
rule .Cmds => .
endmodule

module A

imports A-SYNTAX
imports INT

configuration <T>
  <k> $PGM:Cmds </k>
  <states>
    <state multiplicity="*" type="Map">
      <slot> 0 </slot>
      <data> 0 </data>
    </state>
  </states>
</T>

rule <k> add(K, V) => . ... </k>
     <states>
     (
       .Bag
     =>
       <state>
         <slot> K </slot>
         <data> V </data>
       </state>
     )
       ...
     </states>

endmodule
```

1.a:
```
add(0,0)
add(0,1)
```

Now, krunning `1.a` gets stuck due to the duplicate keys, which is desired:
```
$ krun 1.a
<T>
  <k>
    add ( 0 , 1 ) ~> .Cmds
  </k>
  <states>
    <state>
      <slot>
        0
      </slot>
      <data>
        0
      </data>
    </state>
  </states>
</T>
```

But, kprove "proves" the following spec, which is wrong, because we have no pre-condition that `K` does not appear in other `<state>` cells.
```
module A-SPEC
imports A
         
rule <k> add(K, V) => . ... </k>
     <states>
     (
       .Bag
     =>
       <state>
         <slot> K </slot>
         <data> V </data>
       </state>
     )
       ...
     </states>

endmodule
```

This behavior is confusing, because this example is a minimized version of the full spec reported in https://github.com/kframework/kore/issues/1665, where kprove failed to prove it.

**Comment by daejunpark** [comment]
For what it's worth, here is another confusing behavior for a slightly different spec.

```
module A2-SPEC
imports A
  
rule <k> add(K, V) => . ... </k>
     <states>
       <state>
         <slot> K </slot>
         <data> V </data>
       </state>
     (
       .Bag
     =>  
       <state>
         <slot> K +Int 1 </slot>
         <data> V </data>
       </state>
     )
       ...
     </states>

endmodule
```

Now, kprove still passes the above spec, but outputs a warning.
```
$ kprove a2-spec.k 
kore-exec: [2020-03-12 13:45:50.503629541] Warning (WarnBottomHook): Evaluating 'MAP.concat' in pattern:
    /* Fn Spa */
    Lbl'Unds'StateCellMap'Unds'{}(
        /* Fl Fn D Sfa */
        /* builtin: */
        /* element: */ LblStateCellMapItem{}(
            /* Fl Fn D Sfa */
            Lbl'-LT-'slot'-GT-'{}(/* Fl Fn D Sfa */ VarK:SortInt{}),
            /* Fl Fn D Sfa */
            Lbl'-LT-'state'-GT-'{}(
                /* Fl Fn D Sfa */
                Lbl'-LT-'slot'-GT-'{}(/* Fl Fn D Sfa */ VarK:SortInt{}),
                /* Fl Fn D Sfa */
                Lbl'-LT-'data'-GT-'{}(/* Fl Fn D Sfa */ VarV:SortInt{})
            )
        ),
        /* Fn Sfa */
        /* builtin: */
        Lbl'Unds'StateCellMap'Unds'{}(
            /* element: */ LblStateCellMapItem{}(
                /* Fl Fn D Sfa */
                Lbl'-LT-'slot'-GT-'{}(/* Fl Fn D Sfa */ VarK:SortInt{}),
                /* Fl Fn D Sfa */
                Lbl'-LT-'state'-GT-'{}(
                    /* Fl Fn D Sfa */
                    Lbl'-LT-'slot'-GT-'{}(/* Fl Fn D Sfa */ VarK:SortInt{}),
                    /* Fl Fn D Sfa */
                    Lbl'-LT-'data'-GT-'{}(/* Fl Fn D Sfa */ VarV:SortInt{})
                )
            ),
            /* opaque child: */ /* Fl Fn D Sfa */ VarDotVar3:SortStateCellMap{}
        )
    )
has resulted in \bottom.
#True
```

**Comment by ttuegel** [comment]
Ah! I know what the problem is: it's not actually getting stuck on the branch with duplicate keys. If there are duplicate keys, then the configuration is `\bottom`, and the proof is trivial (`\bottom → _`).

This is actually a duplicate of #1637 !

**Comment by ttuegel** [comment]
> But, kprove "proves" the following spec, which is wrong, because we have no pre-condition that `K` does not appear in other `<state>` cells.

This is not wrong. If `K` _does_ appear in another `<state>` cell, then the configuration is `\bottom`; then the claim is `\bottom → _`, which is trivially true.

The output of `krun` is actually wrong, because the backend doesn't get stuck. Instead, it applies the `add` rule to get the final configuration, which is `\bottom`. When that happens, the backend outputs the last defined configuration. That's not correct, but it's more useful than printing `\bottom` for everything undefined.

**Comment by ttuegel** [comment]
The `A2-SPEC` passes correctly, for the same reason that `A-SPEC` passes.

**Comment by daejunpark** [comment]
I don't get it.

An instance of the LHS of A-SPEC is:
```
<k> add(K, V) </k>
<states>
  <state>
    <slot> K </slot>
    <data> _ </data>
  </state>
</states>
```
which is _not_ bottom, but it gets stuck, and cannot be matched by the RHS, which is a counter-example of the spec, no?

Similarly, a counter-example instance of the LHS of A2-SPEC is:
```
<k> add(K, V) </k>
<states>
  <state>
    <slot> K </slot>
    <data> _ </data>
  </state>
  <state>
    <slot> K +Int 1 </slot>
    <data> _ </data>
  </state>
</states>
```

**Comment by ttuegel** [comment]
Those are not counter-examples. Let's walk through the proof of `A-SPEC`: Initially, we have `claim[0]`:

```
rule <k> add(K, V) => . ... </k>
     <states>
     (
       .Bag
     =>
       <state>
         <slot> K </slot>
         <data> V </data>
       </state>
     )
       ...
     </states>
```

In Kore, this claim has the form `config[0] → _` where `config[0]` is:

```
<k> add(K, V) ... </k> <states>...</states>
```
```

I am leaving off the right-hand side of the initial claim because it's not important. First, we apply this rule:

```
rule <k> add(K, V) => . ... </k>
     <states>
     (
       .Bag
     =>
       <state>
         <slot> K </slot>
         <data> V </data>
       </state>
     )
       ...
     </states>
```

to get `config[1]`:

```
<k> ... </k>
<states>
  <state>
    <slot> K </slot>
    <data> V </data>
  </state>
  ...
</states>
```

Having taken a step, the remaining claim to be proved is `config[1] → _`. If `K` is not in the remainder of the `<states>` (the `...`) then the claim is proven. But if `K` _is_ in the remainder, then `config[1] = \bottom`, and we have `\bottom → _` which is _also_ true.

**Comment by daejunpark** [comment]
I don't think so. The semantics rule (axiom) should not be applied if K is in `...`. So, `config[1]` shouldn't be generated in the first place, if K is in `...`.

Also, think about purely the reachability claim, apart from the underlying haskell backend algorithm.  If K is in `...`, the RHS of the spec is bottom.  How can we prove a spec `_ => bottom`?  It doesn't make sense to me at all.  If you (or the KORE team) have a different interpretation of the reachability claim, I'll strongly argue that it should be fixed.

**Comment by daejunpark** [comment]
OK. As I thought more about it, it seems to be about the semantics (i.e., axioms), not the spec.

In an extreme case, if we have an axiom `rule foo => bottom`, then any spec like `rule foo => _` will be passed by kprove.

Now, in this particular issue, the A module can be further simplified to:
```
module A-MINIMAL
rule foo => bar requires someConstraint
rule foo => bottom [owise]
endmodule
```
and kprove "succeeds" in proving:
```
module A-MINIMAL-SPEC
rule foo => bar
endmodule
```

I think we need a way to _*reject*_ such axioms in the first place. Otherwise, it is very possible that we model a critical system in K, but we make a mistake having the model contains a false axiom, then we can prove properties of the system model that are not satisfied by the system, which is very bad!

**Comment by traiansf** [comment]
we could try implementing some checks, but we shouldn't expect a general solution to this problem.  Proving a set of axioms consistent does not seem like a decidable task.

consider the simple rule `x:Int / Y:Int => X /Int Y`

This rule is generally OK, but if you instantiate it on `y=0`, you get
`x:Int / 0 -> \next (X /Int 0)`

Now you don't immediately know that `X /Int 0` is bottom, as you have no axiom stating that.
Instead we have that `\ceil(@X /Int @Y) = \ceil(@X) /\ \ceil(@Y) /\ \equals(y ==Int 0, false)`, which, if
we instantiate with `@X = x` and `@Y = 0` we obtain `\ceil(x /Int 0) = \ceil(x) /\ \ceil(0) /\ \equals(0 ==Int 0, false)`, which leads to ``\ceil(x /Int 0) = \bottom` because `0 ==Int 0 = true`.
From this we can deduce that `x /Int 0 = \bottom`, whence 
`x:Int / 0 -> \next \bottom`, whence `x:Int / 0 -> \bottom`, i.e., `x:Int / 0 = \bottom`. which makes the specification inconsistent.

I know we all have probably seen this argument already; I was just trying to make a point that it might be harder to automatically detect inconsistencies.


I guess what we could do would be to try and find a substitution such that a RHS of a rule (including ensures) is \bottom and the lhs of the rule (including requires) is not bottom.

**Comment by daejunpark** [comment]
@traiansf Thanks. I didn't ask a perfect solution, but a practical measure to prevent potential mistakes that can likely happen.

> I guess what we could do would be to try and find a substitution such that a RHS of a rule (including ensures) is \bottom and the lhs of the rule (including requires) is not bottom.

I thought a similar thing too. That is, for each `rule l => r requires c`, prove that `\ceil(l /\ c)` implies `\ceil(r /\ c)`. 

BTW, I couldn't imagine a practical example semantics that can generate a bottom, even if all its rules satisfy the above condition.

**Comment by ttuegel** [comment]
Fixed in #1911.
